backend/otel/otel_test.go

```diff
@@ -84,19 +86,38 @@ func TestHandler_HandleTrace(t *testing.T) {
 	}
 	h.HandleTrace(w, r)
 
-	assert.Equal(t, 4, len(producer.messages))
-	assert.Equal(t, kafkaqueue.PushBackendPayload, producer.messages[0].Type)
-	assert.Equal(t, kafkaqueue.MarkBackendSetup, producer.messages[1].Type)
-	if assert.Equal(t, kafkaqueue.PushLogs, producer.messages[2].Type) {
-		assert.Equal(t, 14, len(producer.messages[2].PushLogs.LogRows))
-		for _, log := range producer.messages[2].PushLogs.LogRows {
-			assert.Equal(t, model.LogSourceBackend, log.Source)
-		}
-	}
-	if assert.Equal(t, kafkaqueue.PushLogs, producer.messages[3].Type) {
-		assert.Equal(t, 1, len(producer.messages[3].PushLogs.LogRows))
-		for _, log := range producer.messages[3].PushLogs.LogRows {
-			assert.Equal(t, model.LogSourceFrontend, log.Source)
+	assert.GreaterOrEqual(t, 4, len(producer.messages))
+
+	_, ok := lo.Find(producer.messages, func(message *kafkaqueue.Message) bool {
+		return message.Type == kafkaqueue.PushBackendPayload
+	})
+	assert.Truef(t, ok, "did not find a PushBackendPayload message")
+
+	_, ok = lo.Find(producer.messages, func(message *kafkaqueue.Message) bool {
+		return message.Type == kafkaqueue.MarkBackendSetup
+	})
+	assert.Truef(t, ok, "did not find a MarkBackendSetup message")
+
+	_, ok = lo.Find(producer.messages, func(message *kafkaqueue.Message) bool {
+		return message.Type == kafkaqueue.PushLogs
+	})
+	assert.Truef(t, ok, "did not find a PushLogs message")
+
+	allPushLogs := lo.Filter(producer.messages, func(message *kafkaqueue.Message, _ int) bool {
+		return message.Type == kafkaqueue.PushLogs
+	})
+
+	for _, pushLogs := range allPushLogs {
+		if len(pushLogs.PushLogs.LogRows) == 14 {
```

Comment:
Is the issue for the flaking test that this test runs with different inputs? It feels a bit weird for the test runs to not be consistent. Is there a way to modify this test such that it runs with a consistent output?

Reply:
the test runs with the same inputs but the data is submitted to the mock kafka object in an arbitrary order (no ordering guarantees), so the random order between frontend / backend logs and errors broke this previous logic. 
https://github.com/highlight/highlight/blob/1d0409e596a39955efded4c6c7b4403f15c44dba/backend/otel/otel.go#L182-L185
we use go maps to store the deserialized data before it is sent to kafka. since frontend and backend logs and errors live in the same maps, but maps have no guarantees about iteration order, the [order at kafka submission](https://github.com/highlight/highlight/blob/1d0409e596a39955efded4c6c7b4403f15c44dba/backend/otel/otel.go#L261) is arbitrary, so we end up with about 1/3 of the runs having `PushBackendPayload, PushLogs, MarkBackendSetup` message ordering and the rest having `PushBackendPayload, MarkBackendSetup, PushLogs` ordering.

Reply:
Got it! Thanks for describing that.